File: chatbase/src/hooks/common/utm.ts
/**
 * 유입 출처(UTM) 수집.
 *
 * 주소의 `utm_*`을 잡아 두었다가 **가입 요청에 실어 보낸다.** 서버는 그 값을
 * `tb_users`에 박고, 어드민이 "어느 채널이 결제까지 갔나"를 집계한다.
 *
 * GA4만으로는 안 되는 이유: GA4는 "cafe_apsa에서 34명 방문"까지만 알고
 * 우리 결제 데이터를 모른다. 둘을 잇는 값이 여기서 나온다.
 *
 * 값 규칙과 예시는 `SALES.md` §6.
 */

const KEY = "chatbase_utm";

/**
 * 30일. 반년 전 방문의 꼬리표가 오늘 가입에 붙으면 그건 신호가 아니라 노이즈다.
 */
const MAX_AGE_MS = 30 * 24 * 60 * 60 * 1000;

/** 서버의 `UTM_FIELDS`(app/core/utils/utm.py)와 같아야 한다. */
const FIELDS = ["utm_source", "utm_medium", "utm_campaign"] as const;

/** `tb_users.utm_*` 이 String(100)이다. 서버도 자르지만 보내기 전에 맞춘다. */
const MAX_LEN = 100;

export type Utm = Partial<Record<(typeof FIELDS)[number], string>>;

interface Stored {
  v: Utm;
  /** 저장 시각(ms). 만료 판정용. */
  t: number;
}

/**
 * localStorage 는 시크릿 모드·저장소 차단에서 **접근 자체가 던진다.**
 * 유입 측정 때문에 가입이 막히면 안 되므로 전부 삼키고 없는 셈 친다.
 */
const load = (): Stored | null => {
  try {
    const raw = localStorage.getItem(KEY);
    if (!raw) return null;
    const parsed = JSON.parse(raw) as Stored;
    if (!parsed?.t || Date.now() - parsed.t > MAX_AGE_MS) return null;
    return parsed;
  } catch {
    return null;
  }
};

/**
 * 주소에 `utm_*`이 있으면 저장한다. 앱이 뜰 때 한 번 부른다.
 *
 * **먼저 들어온 것을 유지한다(first-touch).** 알고 싶은 건 "누가 데려왔나"지
 * "마지막에 뭘 눌렀나"가 아니다. 카페 글을 보고 왔다가 며칠 뒤 검색으로 다시
 * 들어와 가입했다면, 공은 카페 글에 있다.
 *
 * `utm_source`가 없으면 아무것도 안 한다 — 집계 기준이 source라서
 * medium·campaign만 있는 건 쓸 데가 없다.
 */
export const captureUtm = (): void => {
  const query = new URLSearchParams(window.location.search);

  const found: Utm = {};
  for (const field of FIELDS) {
    const value = query.get(field);
    if (value) found[field] = value.slice(0, MAX_LEN);
  }

  if (!found.utm_source) return;
  if (load()) return; // first-touch — 이미 있으면 덮지 않는다

  try {
    localStorage.setItem(KEY, JSON.stringify({ v: found, t: Date.now() }));
  } catch {
    // 저장 못 해도 가입은 되어야 한다
  }
};

/**
 * 저장된 값. 없으면 빈 객체라 요청 바디에 그대로 펼쳐 넣어도 안전하다.
 *
 * ```ts
 * mutate({ email, password, ...readUtm() })
 * ```
 */
export const readUtm = (): Utm => load()?.v ?? {};

File: backend/app/core/utils/utm.py
"""유입 출처(UTM) 다루기.

가입 요청 바디로 넘어온 `utm_*`을 골라내 `tb_users`에 박는 데 쓴다.

**왜 저장하는가** — GA4는 "cafe_apsa에서 34명 방문"까지만 안다. 그중 누가 결제까지
갔는지는 우리 DB에만 있고, 둘을 잇는 값이 없으면 어느 카페가 돈이 됐는지 알 수 없다.

**왜 지금인가** — 가입하는 순간에만 알 수 있어서 소급이 안 된다. 나중에 붙이면
그전에 가입한 사람의 출처는 영영 모른다.

규칙과 값 예시는 `chatbase/SALES.md` §6.
"""

#: DB 컬럼과 1:1. 늘리려면 `tb_users` 마이그레이션도 같이 가야 한다.
UTM_FIELDS = ("utm_source", "utm_medium", "utm_campaign")

#: `tb_users.utm_*` 이 String(100)이다. 넘치면 DB가 던지므로 여기서 자른다.
#: 방문자가 주소창에 무엇을 넣든 우리 저장은 안 깨져야 한다.
_MAX_LEN = 100


def extract_utm(body: dict | None) -> dict | None:
    """요청 바디에서 `utm_*`만 골라낸다. 쓸 값이 없으면 None.

    바디는 클라이언트가 보낸 값이라 타입을 믿지 않는다 — 문자열이 아니거나
    공백뿐이면 버린다. 셋 다 없으면 None을 돌려주고, 호출부는 컬럼을 아예
    건드리지 않는다(직접 방문·검색 유입이 여기 해당한다).
    """
    if not isinstance(body, dict):
        return None

    picked = {}
    for field in UTM_FIELDS:
        value = body.get(field)
        if isinstance(value, str) and value.strip():
            picked[field] = value.strip()[:_MAX_LEN]

    # source 없이 medium·campaign만 오는 건 의미가 없다(집계 기준이 source다).
    return picked if picked.get("utm_source") else None


def utm_columns(utm: dict | None) -> dict:
    """`User(**...)`에 펼쳐 넣을 kwargs.

    utm이 없으면 **빈 dict**를 준다 — 컬럼을 명시하지 않으면 NULL로 남는다.
    None을 명시적으로 넣는 것과 결과는 같지만, 호출부에서 "안 건드린다"는
    의도가 더 분명해진다.
    """
    if not utm:
        return {}
    return {field: utm.get(field) for field in UTM_FIELDS}
